packages/backend/src/routes/UsernameRouter.ts

```diff
@@ -0,0 +1,65 @@
+import { Express } from 'express'
+import catchError from '../catchError'
+import { ethers } from 'ethers'
+import {
+    UNIREP,
+    UNIREP_ABI,
+    DEFAULT_ETH_PROVIDER,
+    UNIREP_SOCIAL,
+    UNIREP_SOCIAL_ABI,
+    ADMIN_SESSION_CODE,
+} from '../constants'
+import TransactionManager from '../daemons/TransactionManager'
+import { formatProofForSnarkjsVerification } from '@unirep/circuits'
+import { ReputationProof, BaseProof } from '@unirep/contracts'
+import { hashOne } from '@unirep/crypto'
+
+export default (app: Express) => {
+    app.post('/api/usernames', catchError(setUsername))
+}
+
+async function setUsername(req, res) {
+    const unirepContract = new ethers.Contract(
+        UNIREP,
+        UNIREP_ABI,
+        DEFAULT_ETH_PROVIDER
+    )
+    const unirepSocialContract = new ethers.Contract(
+        UNIREP_SOCIAL,
+        UNIREP_SOCIAL_ABI,
+        DEFAULT_ETH_PROVIDER
+    )
+
+    // accept a requested new username and ZK proof proving the epoch key and current username (graffiti pre-image)
+    const { newUsername, publicSignals, proof } = req.body
+    const usernameProof = new ReputationProof(
```

Comment:
You need to verify this proof and return an error if it's invalid. You also need to check the `globalStateTree` public signal (make sure it exists in the current epoch).

Reply:
You should also check that `proveGraffiti` (another public signal) is `1`.

Reply:
Would users have to have a certain amount of karma in order to make a username?

Reply:
No I think they should be able to claim a name regardless of rep balance. They can only claim 1 username so it seems fine to let negative rep people do it.

The `ReputationProof` circuit is the only thing we have for verifying graffiti pre-image right now. As a separate task we should write a circuit that proves an epoch key and graffiti pre-image only. It will save lots of calldata.

For the `ReputationProof` you should be able to pass a value `minRep: 0` or `proveMinRep: 0` to prove nothing about the reputation.

Reply:
gotcha. I will raise this as a new issue.
> As a separate task we should write a circuit that proves an epoch key and graffiti pre-image only. It will save lots of calldata.